modules/samples/virtual_firealarm/src/org.wso2.carbon.device.mgt.iot.sample.virtual.firealarm.service.impl/src/main/java/org/wso2/carbon/device/mgt/iot/sample/virtual/firealarm/service/impl/VirtualFireAlarmService.java

```diff
@@ -601,40 +605,40 @@ public String requestSonarReading(@HeaderParam("owner") String owner,
 
 	@Path("controller/readtemperature")
 	@GET
-	public String requestTemperature(@HeaderParam("owner") String owner,
+    @Consumes("application/json")
+    @Produces("application/json")
+	public TemperatureRecord requestTemperature(@HeaderParam("owner") String owner,
 	                                 @HeaderParam("deviceId") String deviceId,
 	                                 @HeaderParam("protocol") String protocol,
 	                                 @Context HttpServletResponse response) {
-		String replyMsg = "";
+        TemperatureRecord temperatureRecord = null;
 
 		DeviceValidator deviceValidator = new DeviceValidator();
 		try {
 			if (!deviceValidator.isExist(owner, SUPER_TENANT, new DeviceIdentifier(deviceId, VirtualFireAlarmConstants.DEVICE_TYPE))) {
 				response.setStatus(Response.Status.UNAUTHORIZED.getStatusCode());
-				return "Unauthorized Access";
 			}
 		} catch (DeviceManagementException e) {
-			replyMsg = e.getErrorMessage();
 			response.setStatus(Response.Status.INTERNAL_SERVER_ERROR.getStatusCode());
```

Comment:
put appropriate error log messages as a way of conveying an error has occurred at server-side.